Bug: test_demo_actor flake

[flake seen here.](https://github.com/dapr/python-sdk/actions/runs/26515802197/job/78092497418?pr=1060)
```Run uv run pytest tests/examples/
============================= test session starts ==============================
platform linux -- Python 3.11.15, pytest-9.0.2, pluggy-1.6.0
rootdir: /home/runner/work/python-sdk/python-sdk
configfile: pyproject.toml
plugins: langsmith-0.7.17, asyncio-1.3.0, anyio-4.12.1
asyncio: mode=Mode.AUTO, debug=False, asyncio_default_fixture_loop_scope=None, asyncio_default_test_loop_scope=function
collected 32 items

tests/examples/test_configuration.py .                                   [  3%]
tests/examples/test_conversation.py ..                                   [  9%]
tests/examples/test_crypto.py ..                                         [ 15%]
tests/examples/test_demo_actor.py F                                      [ 18%]
tests/examples/test_distributed_lock.py .                                [ 21%]
tests/examples/test_error_handling.py .                                  [ 25%]
tests/examples/test_grpc_proxying.py .                                   [ 28%]
tests/examples/test_invoke_binding.py .                                  [ 31%]
tests/examples/test_invoke_custom_data.py .                              [ 34%]
tests/examples/test_invoke_http.py .                                     [ 37%]
tests/examples/test_invoke_simple.py .                                   [ 40%]
tests/examples/test_jobs.py ..                                           [ 46%]
tests/examples/test_langgraph_checkpointer.py .                          [ 50%]
tests/examples/test_mcp.py .                                             [ 53%]
tests/examples/test_metadata.py .                                        [ 56%]
tests/examples/test_pubsub_simple.py .                                   [ 59%]
tests/examples/test_pubsub_streaming.py ..                               [ 65%]
tests/examples/test_pubsub_streaming_async.py ..                         [ 71%]
tests/examples/test_secret_store.py ..                                   [ 78%]
tests/examples/test_state_store.py .                                     [ 81%]
tests/examples/test_state_store_query.py .                               [ 84%]
tests/examples/test_w3c_tracing.py .                                     [ 87%]
tests/examples/test_workflow.py ....                                     [100%]

=================================== FAILURES ===================================
_______________________________ test_demo_actor ________________________________

dapr = <conftest.DaprRunner object at 0x7f84b5f31010>

    @pytest.mark.example_dir('demo_actor/demo_actor')
    def test_demo_actor(dapr):
        dapr.start(
            '--app-id demo-actor --app-port 3000 -- uvicorn --port 3000 demo_actor_service:app',
            wait=10,
        )
        client_output = dapr.run(
            '--app-id demo-client -- python3 demo_actor_client.py',
            timeout=60,
        )
        for line in EXPECTED_CLIENT:
>           assert line in client_output, f'Missing in client output: {line}'
E           AssertionError: Missing in client output: null
E           assert 'null' in 'ℹ️  Starting Dapr with id demo-client. HTTP Port: 39753. gRPC Port: 36149\nℹ️  Checking if Dapr sidecar is listening ...rminated signal received: shutting down\n✅  Exited Dapr successfully\n❌  Error exiting App: app exited with status 1\n'

tests/examples/test_demo_actor.py:41: AssertionError
----------------------------- Captured stdout call -----------------------------
ℹ️  Starting Dapr with id demo-client. HTTP Port: 39753. gRPC Port: 36149
ℹ️  Checking if Dapr sidecar is listening on HTTP port 39753
Flag --components-path has been deprecated, use --resources-path
time="2026-05-27T14:01:03.826645598Z" level=info msg="Starting Dapr Runtime -- version 1.18.0-rc.4 -- commit 8c7c094049c1697978cec5701c2b38d489360f0e" app_id=demo-client instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.826663742Z" level=info msg="Log level set to: info" app_id=demo-client instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.826799358Z" level=warning msg="mTLS is disabled. Skipping certificate request and tls validation" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.security type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.827053152Z" level=info msg="Enabled features: HotReload WorkflowsRemoteActivityReminder" app_id=demo-client instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.827104578Z" level=info msg="metric spec: {\"enabled\":true}" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.diagnostics type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.827124154Z" level=info msg="Using default latency distribution buckets: [1 2 3 4 5 6 8 10 13 16 20 25 30 40 50 65 80 100 130 160 200 250 300 400 500 650 800 1000 2000 5000 10000 20000 50000 100000]" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.diagnostics type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.827192642Z" level=warning msg="The default value for 'spec.metric.http.increasedCardinality' will change to 'false' in Dapr 1.15 or later" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.diagnostics type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.827487983Z" level=info msg="Watching directories: [/home/runner/.dapr/components]" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.hotreload.loader.disk type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.827705245Z" level=info msg="standalone mode configured" app_id=demo-client instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.827723439Z" level=info msg="app id: demo-client" app_id=demo-client instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.827811293Z" level=info msg="Dapr trace sampler initialized: ParentBased{root:TraceIDRatioBased{1},remoteParentSampled:AlwaysOnSampler,remoteParentNotSampled:AlwaysOffSampler,localParentSampled:AlwaysOnSampler,localParentNotSampled:AlwaysOffSampler}" app_id=demo-client instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.827927453Z" level=info msg="metrics server started on 0.0.0.0:38405/" app_id=demo-client instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.828335731Z" level=info msg="local service entry announced: demo-client -> 10.1.0.174:43307" app_id=demo-client component="nr (mdns/v1)" instance=runnervmg397c scope=dapr.contrib type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.828402115Z" level=info msg="Initialized name resolution to mdns" app_id=demo-client instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.828503815Z" level=info msg="Loading components…" app_id=demo-client instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.829467132Z" level=info msg="Connected and received scheduler hosts addresses: [localhost:50006]" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.scheduler.watchhosts type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.829613695Z" level=info msg="Scheduler client initialized for address: localhost:50006" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.scheduler.clients type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.829635175Z" level=info msg="Scheduler clients initialized" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.scheduler.clients type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.831516889Z" level=info msg="Component loaded: pubsub (pubsub.redis/v1)" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.processor type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.831636639Z" level=info msg="Waiting for all outstanding components to be processed…" app_id=demo-client instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.832800093Z" level=info msg="Using 'statestore' as actor state store" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.processor.state type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.832824919Z" level=info msg="Component loaded: statestore (state.redis/v1)" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.processor type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.832834417Z" level=info msg="All outstanding components processed" app_id=demo-client instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.832840358Z" level=info msg="Loading endpoints…" app_id=demo-client instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.833049107Z" level=info msg="Waiting for all outstanding http endpoints to be processed…" app_id=demo-client instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.833067011Z" level=info msg="All outstanding http endpoints processed" app_id=demo-client instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.833073022Z" level=info msg="Loading Declarative Subscriptions…" app_id=demo-client instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.833288093Z" level=warning msg="App channel is not initialized. Did you configure an app-port?" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.channels type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.833460264Z" level=info msg="gRPC server listening on TCP address: :36149" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.grpc.api type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.83347403Z" level=info msg="Enabled gRPC tracing middleware" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.grpc.api type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.833479069Z" level=info msg="Enabled gRPC metrics middleware" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.grpc.api type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.833525296Z" level=info msg="Registering workflow engine for gRPC endpoint: [::]:36149" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.grpc.api type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.833540674Z" level=info msg="API gRPC server is running on port 36149" app_id=demo-client instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.833551534Z" level=warning msg="The default value for 'spec.metric.http.increasedCardinality' will change to 'false' in Dapr 1.15 or later" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.http type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.833589235Z" level=info msg="Enabled max body size HTTP middleware with size 4194304 bytes" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.http type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.833571759Z" level=info msg="Hot reloading enabled. Daprd will reload 'Component', 'Subscription', 'MCPServer', 'Configuration', 'HTTPEndpoint', 'Resiliency' and 'WorkflowAccessPolicy' resources when they are added, updated or deleted." app_id=demo-client instance=runnervmg397c scope=dapr.runtime.hotreload type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.833600265Z" level=info msg="Enabled tracing HTTP middleware" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.http type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.833604804Z" level=info msg="Enabled metrics HTTP middleware" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.http type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.833621802Z" level=info msg="Starting to watch WorkflowAccessPolicy updates" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.hotreload.reconciler type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.833664161Z" level=info msg="Starting to watch Component updates" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.hotreload.reconciler type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.83369005Z" level=info msg="Starting to watch Subscription updates" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.hotreload.reconciler type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.833719254Z" level=info msg="Starting to watch MCPServer updates" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.hotreload.reconciler type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.833746986Z" level=info msg="Starting to watch Configuration updates for SIGHUP reload" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.hotreload.reconciler type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.833757014Z" level=info msg="Starting to watch HTTPEndpoint updates for SIGHUP reload" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.hotreload.reconciler type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.833781851Z" level=info msg="Starting to watch Resiliency updates for SIGHUP reload" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.hotreload.reconciler type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.833855351Z" level=info msg="HTTP server listening on TCP address: :39753" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.http type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.833889705Z" level=info msg="HTTP server is running on port 39753" app_id=demo-client instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.833900886Z" level=info msg="The request body size parameter is: 4194304 bytes" app_id=demo-client instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.833939789Z" level=info msg="gRPC server listening on TCP address: :43307" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.grpc.internal type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.83395087Z" level=info msg="Enabled gRPC tracing middleware" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.grpc.internal type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.833955167Z" level=info msg="Enabled gRPC metrics middleware" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.grpc.internal type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.833971899Z" level=info msg="Internal gRPC server is running on :43307" app_id=demo-client instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.834017864Z" level=info msg="Using Scheduler service for reminders." app_id=demo-client instance=runnervmg397c scope=dapr.runtime.actor.reminders.scheduler type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.834046989Z" level=info msg="Loading MCP servers…" app_id=demo-client instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.83419741Z" level=info msg="Waiting for all outstanding MCP servers to be processed…" app_id=demo-client instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.834216746Z" level=info msg="All outstanding MCP servers processed" app_id=demo-client instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.834226601Z" level=info msg="Actor runtime started" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.actor type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.834229519Z" level=info msg="dapr initialized. Status: Running. Init Elapsed 6ms" app_id=demo-client instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.834237411Z" level=warning msg="Graceful shutdown timeout is infinite, will wait indefinitely to shutdown" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.actor type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.834284769Z" level=info msg="worker started with backend dapr.actors/v1" app_id=demo-client instance=runnervmg397c scope=dapr.wfengine.durabletask.backend type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.834303554Z" level=info msg="Workflow engine started" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.wfengine type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.834388907Z" level=info msg="Scheduler stream connected for [JOB_TARGET_TYPE_JOB]" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.scheduler.cluster type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.834410276Z" level=info msg="Scheduler stream connected for [JOB_TARGET_TYPE_JOB]" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.scheduler.cluster type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.834448164Z" level=info msg="Connected to placement localhost:50005" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.actors.placement.manager.connector.static type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.834506566Z" level=info msg="Scheduler stream connected for [JOB_TARGET_TYPE_JOB]" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.scheduler.cluster type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.835350976Z" level=info msg="Connected to placement service: localhost:50005" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.actors.placement.loops.placement type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.835386362Z" level=info msg="Reporting initial host to placement service with initial types []" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.actors.placement.loops.placement type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.83604357Z" level=info msg="Dissemination complete for version 0 (changed types []), unlocking disseminator default/demo-client" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.actors.placement.loops.disseminator type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.836120844Z" level=info msg="Scheduler stream disconnected" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.scheduler.cluster type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.836124485Z" level=info msg="Scheduler stream disconnected" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.scheduler.cluster type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.836129995Z" level=info msg="Scheduler stream disconnected" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.scheduler.cluster type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.836200914Z" level=info msg="Scheduler stream connected for [JOB_TARGET_TYPE_JOB]" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.scheduler.cluster type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.836244585Z" level=info msg="Scheduler stream connected for [JOB_TARGET_TYPE_JOB]" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.scheduler.cluster type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.836358137Z" level=info msg="Scheduler stream connected for [JOB_TARGET_TYPE_JOB]" app_id=demo-client instance=runnervmg397c scope=dapr.runtime.scheduler.cluster type=log ver=1.18.0-rc.4
ℹ️  Checking if Dapr sidecar is listening on GRPC port 36149
ℹ️  Dapr sidecar is up and running.
ℹ️  Updating metadata for appPID: 5798
ℹ️  Updating metadata for app command: python3 demo_actor_client.py
✅  You're up and running! Both Dapr and your app logs will appear here.

call actor method via proxy.invoke_method()
time="2026-05-27T14:01:05.07352254Z" level=warning msg="Error processing operation DaprBuiltInActorNotFoundRetries. Retrying in 1s…" app_id=demo-client instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:11.078672985Z" level=warning msg="Error processing operation DaprBuiltInActorNotFoundRetries. Retrying in 1s…" app_id=demo-client instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:17.584828258Z" level=warning msg="Error processing operation DaprBuiltInActorNotFoundRetries. Retrying in 1s…" app_id=demo-client instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
Request attempt 1
Sleeping for 1.0 seconds before retrying call
Request attempt 2
Sleeping for 1.5 seconds before retrying call
Request attempt 3
Traceback (most recent call last):
  File "/home/runner/work/python-sdk/python-sdk/examples/demo_actor/demo_actor/demo_actor_client.py", line 81, in <module>
    asyncio.run(main())
  File "/opt/hostedtoolcache/Python/3.11.15/x64/lib/python3.11/asyncio/runners.py", line 190, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.15/x64/lib/python3.11/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.15/x64/lib/python3.11/asyncio/base_events.py", line 654, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/home/runner/work/python-sdk/python-sdk/examples/demo_actor/demo_actor/demo_actor_client.py", line 31, in main
    rtn_bytes = await proxy.invoke_method('GetMyData')
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/python-sdk/python-sdk/dapr/actor/client/proxy.py", line 186, in invoke_method
    return await self._dapr_client.invoke_method(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/python-sdk/python-sdk/dapr/clients/http/dapr_actor_http_client.py", line 74, in invoke_method
    body, _ = await self._client.send_bytes(method='POST', url=url, data=data, headers=headers)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/python-sdk/python-sdk/dapr/clients/http/client.py", line 102, in send_bytes
    raise (await self.convert_to_error(r))
dapr.clients.exceptions.DaprHttpError: error invoke actor method: failed to lookup actor: api error: code = FailedPrecondition desc = did not find address for actor 'DemoActor/1' (Error Code: ERR_ACTOR_INVOKE_METHOD, Status Code: 500)
❌  The App process exited with error code: 1
ℹ️  
terminated signal received: shutting down
✅  Exited Dapr successfully
❌  Error exiting App: app exited with status 1
--------------------------- Captured stdout teardown ---------------------------
ℹ️  Starting Dapr with id demo-actor. HTTP Port: 35051. gRPC Port: 38779
Flag --components-path has been deprecated, use --resources-path
time="2026-05-27T14:00:53.824387098Z" level=info msg="Starting Dapr Runtime -- version 1.18.0-rc.4 -- commit 8c7c094049c1697978cec5701c2b38d489360f0e" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.824402928Z" level=info msg="Log level set to: info" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.824543849Z" level=warning msg="mTLS is disabled. Skipping certificate request and tls validation" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime.security type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.824769159Z" level=info msg="Enabled features: HotReload WorkflowsRemoteActivityReminder" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.824819984Z" level=info msg="metric spec: {\"enabled\":true}" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime.diagnostics type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.824845542Z" level=info msg="Using default latency distribution buckets: [1 2 3 4 5 6 8 10 13 16 20 25 30 40 50 65 80 100 130 160 200 250 300 400 500 650 800 1000 2000 5000 10000 20000 50000 100000]" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime.diagnostics type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.824944857Z" level=warning msg="The default value for 'spec.metric.http.increasedCardinality' will change to 'false' in Dapr 1.15 or later" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime.diagnostics type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.825219499Z" level=info msg="Watching directories: [/home/runner/.dapr/components]" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime.hotreload.loader.disk type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.825672013Z" level=info msg="metrics server started on 0.0.0.0:45067/" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.825747825Z" level=info msg="standalone mode configured" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.82575572Z" level=info msg="app id: demo-actor" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.825817705Z" level=info msg="Dapr trace sampler initialized: ParentBased{root:TraceIDRatioBased{1},remoteParentSampled:AlwaysOnSampler,remoteParentNotSampled:AlwaysOffSampler,localParentSampled:AlwaysOnSampler,localParentNotSampled:AlwaysOffSampler}" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.826191633Z" level=info msg="local service entry announced: demo-actor -> 10.1.0.174:38779" app_id=demo-actor component="nr (mdns/v1)" instance=runnervmg397c scope=dapr.contrib type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.826210598Z" level=info msg="Initialized name resolution to mdns" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.826254671Z" level=info msg="Loading components…" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.827946278Z" level=info msg="Component loaded: pubsub (pubsub.redis/v1)" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime.processor type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.828089914Z" level=info msg="Waiting for all outstanding components to be processed…" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.828269631Z" level=info msg="Connected and received scheduler hosts addresses: [localhost:50006]" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime.scheduler.watchhosts type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.828421573Z" level=info msg="Scheduler client initialized for address: localhost:50006" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime.scheduler.clients type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.828448172Z" level=info msg="Scheduler clients initialized" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime.scheduler.clients type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.83339313Z" level=info msg="Using 'statestore' as actor state store" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime.processor.state type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.833435088Z" level=info msg="Component loaded: statestore (state.redis/v1)" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime.processor type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.833459964Z" level=info msg="All outstanding components processed" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.833471926Z" level=info msg="Loading endpoints…" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.83378477Z" level=info msg="Waiting for all outstanding http endpoints to be processed…" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.833814375Z" level=info msg="All outstanding http endpoints processed" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.833831067Z" level=info msg="Loading Declarative Subscriptions…" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.835635493Z" level=info msg="gRPC server listening on TCP address: :38779" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime.grpc.api type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.835659498Z" level=info msg="Enabled gRPC tracing middleware" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime.grpc.api type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.835643367Z" level=info msg="Hot reloading enabled. Daprd will reload 'Component', 'Subscription', 'MCPServer', 'Configuration', 'HTTPEndpoint', 'Resiliency' and 'WorkflowAccessPolicy' resources when they are added, updated or deleted." app_id=demo-actor instance=runnervmg397c scope=dapr.runtime.hotreload type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.835710683Z" level=info msg="Starting to watch WorkflowAccessPolicy updates" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime.hotreload.reconciler type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.835725731Z" level=info msg="Starting to watch Subscription updates" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime.hotreload.reconciler type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.835670358Z" level=info msg="Enabled gRPC metrics middleware" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime.grpc.api type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.835839934Z" level=info msg="Registering workflow engine for gRPC endpoint: [::]:38779" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime.grpc.api type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.835862807Z" level=info msg="API gRPC server is running on port 38779" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.835882333Z" level=warning msg="The default value for 'spec.metric.http.increasedCardinality' will change to 'false' in Dapr 1.15 or later" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime.http type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.835739577Z" level=info msg="Starting to watch MCPServer updates" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime.hotreload.reconciler type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.835755496Z" level=info msg="Starting to watch Configuration updates for SIGHUP reload" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime.hotreload.reconciler type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.835763872Z" level=info msg="Starting to watch HTTPEndpoint updates for SIGHUP reload" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime.hotreload.reconciler type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.835774051Z" level=info msg="Starting to watch Resiliency updates for SIGHUP reload" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime.hotreload.reconciler type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.835947976Z" level=info msg="Enabled max body size HTTP middleware with size 4194304 bytes" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime.http type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.835792155Z" level=info msg="Starting to watch Component updates" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime.hotreload.reconciler type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.835960188Z" level=info msg="Enabled tracing HTTP middleware" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime.http type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.835968875Z" level=info msg="Enabled metrics HTTP middleware" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime.http type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.836303889Z" level=info msg="HTTP server listening on TCP address: :35051" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime.http type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.836335488Z" level=info msg="HTTP server is running on port 35051" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:00:53.83634717Z" level=info msg="The request body size parameter is: 4194304 bytes" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
INFO:     Started server process [5701]
INFO:     Waiting for application startup.
ℹ️  Updating metadata for appPID: 5701
ℹ️  Updating metadata for app command: uvicorn --port 3000 demo_actor_service:app
✅  You're up and running! Both Dapr and your app logs will appear here.

time="2026-05-27T14:01:03.666246932Z" level=error msg="Failed to listen for gRPC server on TCP address :38779 with error: listen tcp :38779: bind: address already in use" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime.grpc.internal type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.667052295Z" level=info msg="Closing HTTP server [::]:35051…" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime.http type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.667093291Z" level=info msg="stopping mDNS server for app id: demo-actor" app_id=demo-actor component="nr (mdns/v1)" instance=runnervmg397c scope=dapr.contrib type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.667091506Z" level=info msg="Dapr is shutting down" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.66712623Z" level=info msg="Shutting down component statestore (state.redis/v1)" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.667207394Z" level=info msg="Shutting down component pubsub (pubsub.redis/v1)" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.667277364Z" level=info msg="Dapr runtime stopped" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
time="2026-05-27T14:01:03.667556605Z" level=fatal msg="Fatal error from runtime: failed to start internal gRPC server: could not listen on any endpoint" app_id=demo-actor instance=runnervmg397c scope=dapr.runtime type=log ver=1.18.0-rc.4
❌  The daprd process exited with error code: exit status 1
ℹ️  
terminated signal received: shutting down
❌  Error exiting Dapr: exit status 1
✅  Exited App successfully
=========================== short test summary info ============================
FAILED tests/examples/test_demo_actor.py::test_demo_actor - AssertionError: Missing in client output: null
assert 'null' in 'ℹ️  Starting Dapr with id demo-client. HTTP Port: 39753. gRPC Port: 36149\nℹ️  Checking if Dapr sidecar is listening ...rminated signal received: shutting down\n✅  Exited Dapr successfully\n❌  Error exiting App: app exited with status 1\n'
=================== 1 failed, 31 passed in 311.82s (0:05:11) ===================
Error: Process completed with exit code 1.
0s
0s
```

--- a/tests/examples/conftest.py
+++ b/tests/examples/conftest.py
@@ -113,25 +113,61 @@ class DaprRunner:
 
         return ''.join(lines)
 
-    def start(self, args: str, *, wait: int = 5) -> None:
+    def start(self, args: str, *, wait: int = 5, port_bind_retries: int = 3) -> None:
         """Start a long-lived background service.
 
         Use this for servers/subscribers that must stay alive while a second
         process runs via ``run()``. Call ``stop()`` to terminate and collect
         output. Stdout is written to a temp file to avoid pipe-buffer deadlocks.
+
+        Args:
+            args: Arguments passed to ``dapr run``.
+            wait: Seconds to wait for the sidecar to come up before returning.
+            port_bind_retries: Retry count for Dapr sidecar startup failures
+                caused by a transient random-port collision.
         """
-        output_file = tempfile.NamedTemporaryFile(mode='w+', suffix='.log')
-        proc = subprocess.Popen(
-            args=('dapr', 'run', *shlex.split(args)),
-            cwd=self._cwd,
-            stdout=output_file,
-            stderr=subprocess.STDOUT,
-            text=True,
-            **get_kwargs_for_process_group(),
-        )
-        self._bg_process = proc
-        self._bg_output_file = output_file
-        time.sleep(wait)
+        attempts = max(1, port_bind_retries + 1)
+        for attempt in range(attempts):
+            output_file = tempfile.NamedTemporaryFile(mode='w+', suffix='.log')
+            proc = subprocess.Popen(
+                args=('dapr', 'run', *shlex.split(args)),
+                cwd=self._cwd,
+                stdout=output_file,
+                stderr=subprocess.STDOUT,
+                text=True,
+                **get_kwargs_for_process_group(),
+            )
+            time.sleep(wait)
+
+            can_retry = attempt < attempts - 1
+            if can_retry and self._started_with_port_bind_failure(proc, output_file):
+                self._terminate(proc)
+                output_file.close()
+                print(
+                    'Dapr background sidecar failed to bind a random port; '
+                    f'retrying startup after {2**attempt}s '
+                    f'(attempt {attempt + 1}/{attempts})',
+                    flush=True,
+                )
+                time.sleep(2**attempt)
+                continue
+
+            self._bg_process = proc
+            self._bg_output_file = output_file
+            return
+
+    def _started_with_port_bind_failure(
+        self, proc: subprocess.Popen[str], output_file: IO[str]
+    ) -> bool:
+        """Whether a background sidecar already exited from a random-port collision.
+
+        Reads the log only once the process has exited; seeking the temp file
+        while daprd is still writing would corrupt its shared append offset.
+        """
+        if proc.poll() is None:
+            return False
+        output_file.seek(0)
+        return self._is_dapr_port_bind_failure(output_file.read())
 
     def stop(self) -> str:
         """Stop the background service and return its captured output."""

--- a/tests/examples/test_dapr_runner.py
+++ b/tests/examples/test_dapr_runner.py
@@ -1,6 +1,7 @@
 import subprocess
 import time
 from pathlib import Path
+from typing import IO
 
 import pytest
 
@@ -17,6 +18,26 @@ class FakeProcess:
 
     def wait(self, timeout: int | None = None) -> int:
         return self.returncode
+
+
+class FakeBackgroundProcess:
+    """Stand-in for a ``dapr run`` background process started via ``start()``.
+
+    A real background sidecar writes to the file object passed as ``stdout`` and
+    keeps running (``poll()`` returns ``None``); a sidecar that died on a port
+    bind has exited (``poll()`` returns a non-``None`` code).
+    """
+
+    def __init__(self, output: str, returncode: int | None, stdout: IO[str]) -> None:
+        stdout.write(output)
+        stdout.flush()
+        self._returncode = returncode
+
+    def poll(self) -> int | None:
+        return self._returncode
+
+    def wait(self, timeout: int | None = None) -> int:
+        return self._returncode or 0
 
 
 def test_run_retries_transient_dapr_port_bind_failure(
@@ -64,4 +85,54 @@ def test_run_does_not_retry_non_port_bind_failure(monkeypatch, tmp_path: Path) -
     output = DaprRunner(tmp_path).run('--app-id=secretsapp -- python3 example.py', timeout=1)
 
     assert output == 'application failed before printing expected output\n'
+    assert len(popen_calls) == 1
+
+
+PORT_BIND_FAILURE_OUTPUT = (
+    'level=error msg="Failed to listen for gRPC server on TCP address :38779 '
+    'with error: listen tcp :38779: bind: address already in use"\n'
+    'level=fatal msg="Fatal error from runtime: failed to start internal gRPC '
+    'server: could not listen on any endpoint"\n'
+)
+
+
+def test_start_retries_transient_dapr_port_bind_failure(
+    monkeypatch, tmp_path: Path, capsys: pytest.CaptureFixture[str]
+) -> None:
+    attempts = [
+        (PORT_BIND_FAILURE_OUTPUT, 1),
+        ('INFO:     Application startup complete.\n', None),
+    ]
+    popen_calls = []
+
+    def fake_popen(*args, **kwargs) -> FakeBackgroundProcess:
+        popen_calls.append((args, kwargs))
+        output, returncode = attempts.pop(0)
+        return FakeBackgroundProcess(output, returncode, kwargs['stdout'])
+
+    monkeypatch.setattr(subprocess, 'Popen', fake_popen)
+    sleeps: list[int] = []
+    monkeypatch.setattr(time, 'sleep', sleeps.append)
+
+    DaprRunner(tmp_path).start('--app-id demo-actor -- uvicorn demo:app', wait=0)
+
+    assert len(popen_calls) == 2
+    assert sleeps == [0, 1, 0]
+    assert (
+        'Dapr background sidecar failed to bind a random port; retrying startup after 1s'
+        in capsys.readouterr().out
+    )
+
+
+def test_start_does_not_retry_non_port_bind_failure(monkeypatch, tmp_path: Path) -> None:
+    popen_calls = []
+
+    def fake_popen(*args, **kwargs) -> FakeBackgroundProcess:
+        popen_calls.append((args, kwargs))
+        return FakeBackgroundProcess('app crashed for an unrelated reason\n', 1, kwargs['stdout'])
+
+    monkeypatch.setattr(subprocess, 'Popen', fake_popen)
+
+    DaprRunner(tmp_path).start('--app-id demo-actor -- uvicorn demo:app', wait=0)
+
     assert len(popen_calls) == 1

